### PR TITLE
Implement contention via lock for the selectors in ConnectionManager.

### PR DESCRIPTION
**Pull request by cyraxjoe**
❓ **What kind of change does this PR introduce?**
  - [x] 🐞 bug fix
  - [ ] 🐣 feature
  - [ ] 📋 docs update
  - [x] 📋 tests/coverage improvement
  - [x] 📋 refactoring
  - [ ] 💥 other

📋 **What is the related issue number (starting with `#`)**

#317 

❓ **What is the current behavior?** (You can also link to an open issue here)

There is a race condition between the internal server `tick()` method to expire connections and the
worker threads that can modify the underlying sockets to be selected, while is been iterating over
by the expire method of `ConnectionManager`.

❓ **What is the new behavior (if this is a feature change)?**
Limits the access methods that could  modify the underlying sockets/fd that are been selected by the
`selectors.DefaultSelector` by having a write lock. An indirection layer that will guarantee consistency
between thread at the cost of a little overhead.

📋 **Other information**:

This locking approach could be temporary in case that we refactor the whole use of `selectors`, `ConnectionManager`
and `Server`.

Additionally: 
 - this PR expands the use of the `ErrorLogManager` helper test class, now is also used by `test_wsgi`, this test module
seems to be extensive enough to trigger some of the errors that were reported in issue #60 
- remove the `xfail` mark in one test that seems that it wasn't failing after all, it just wasn't returning the right value the test wsgi app.
- remove the py2 exception to tolerate the serve error_log messages, it was covering an issue in the different ways that `Server` closes the connections, for py2 it closed directly the OS socket, given the comment that was in that segment (from around 12 years)  seems that the goal was to send the FIN packet, which can be accomplished by socket.shutdown, instead of relying in the OS by closing the file descriptor. (verified with linux/wireshark/pdb at localhost).

📋 **Checklist**:

  - [X] I think the code is well written
  - [X] I wrote [good commit messages][1]
  - [ ] I have [squashed related commits together][2] after the changes have been approved
  - [ ] Unit tests for the changes exist
  - [X] Integration tests for the changes exist (if applicable)
  - [X] I used the same coding conventions as the rest of the project
  - [X] The new code doesn't generate linter offenses
  - [ ] Documentation reflects the changes
  - [ ] The PR relates to *only* one subject with a clear title
  and description in grammatically correct, complete sentences

[1]: http://chris.beams.io/posts/git-commit/
[2]: https://github.com/todotxt/todo.txt-android/wiki/Squash-All-Commits-Related-to-a-Single-Issue-into-a-Single-Commit

---
This change is [Reviewable](https://reviewable.io/reviews/cherrypy/cheroot/318)

### CODE REVIEW

File: cheroot/test/test_conn.py [hunk 1]
```diff
@@ -1013,16 +1013,11 @@ def test_No_CRLF(test_client, invalid_terminator):
 
 
 class FaultySelectorManager:
-    """Mock class to insert errors in th execution.
-
-    Wraps: :py:class:`!cheroot.connections._SelectorManager`.
```
Comment (webknjaz) [inline cheroot/test/test_conn.py]: @cyraxjoe FYI look how I've ignored similar references to undocumented objects in the past: https://github.com/cherrypy/cheroot/blob/d2f28f2/docs/conf.py#L95-L96

File: cheroot/test/test_conn.py [hunk 2]
```diff
@@ -72,7 +68,7 @@ def err_before_read(req, resp):
 
     def one_megabyte_of_a(req, resp):
         """Render 1MB response."""
-        return ['a' * 1024] * 1024
+        return 'a' * 1024 * 1024
```
Comment (webknjaz) [inline cheroot/test/test_conn.py:71]: Why did you change a series of chunks with one blob?

File: cheroot/test/test_conn.py [hunk 3]
```diff
@@ -1017,12 +963,11 @@ def test_Content_Length_out(
     ))
 
 
-@pytest.mark.xfail(
```
Comment (webknjaz) [inline cheroot/test/test_conn.py:1020]: @cyraxjoe this was failing but flakily. I was unable to reproduce this on every run.

**Comment by webknjaz** [comment]
@cyraxjoe I want to explore #319 for solving this first. Do you think you could extract tests into a separate PR so that they could be merged earlier? Plus, it'd make this one easier to review.

File: cheroot/test/test_conn.py (hunk 2 above)
Comment (cyraxjoe) [reply cheroot/test/test_conn.py:71]: I think that I did this because of the way that the  helper `Controller` is implemented https://github.com/cherrypy/cheroot/blob/master/cheroot/test/helper.py#L159-L169, it doesn't calculate the right size for the response if the handler didn't specified the `Content-Length` header. I remember that I was looking over the traffic with wireshark and saw that the response was there but the application only read the first 1024 bytes because of the `Content-Length` header, the rest was discarded. I'll double check then, I think my "fix" was only fixing the integration with testing code.

**Comment by cyraxjoe** [comment]
@webknjaz  I think that only the `ErrorLogManitor` refactor could be helpful in the other PR (to include the verification in the test  in `test_wsgi.py`). The other test is _heavily_ dependent on the final implementation, in this case is tightly coupled with the `_SelectorManager`, @liamstask  would need to verify the triggers of the errors paths to trigger those and verify the output in the coverage report.
